design/sync-labels-bmh-to-node.md

```diff
@@ -0,0 +1,191 @@
+
+### New API Type
+
+We could introduce a new API type: `BareMetalHostLabelSyncProfile` (any other suggestions for the name?).
+
+The common case is when a label on the BMH is updated, we must first find the corresponding `BareMetalHostLabelSyncProfile` (if any) that matches the BMH. To avoid fetching all `BareMetalHostLabelSyncProfile`, we can make use of a label on the BMH, in namespace/name format, that references the `BareMetalHostLabelSyncProfile` (see example below). We can drop the namespace requirement if it's in the same namespace as the BMH.
+
+The less common case is where the `BareMetalHostLabelSyncProfile` is updated (e.g. a prefix is removed). In this case, we will need to fetch all BMHs associated and perform a sync operations across the impacted Nodes. Label selectors could be used for this purpose. 
+
+The approach is illustrated below:
+
+```
+kind: BareMetalHost
+metadata:
+  labels:
+    metal3.io/baremetalhost-label-sync-profile: some-ns/label-sync-profile-0
+...
+---
+kind: BareMetalHostLabelSyncProfile
+name: label-sync-profile-0
+namespace: some-ns
+labelSelector: 
+  matchLabels: 
+    metal3.io/baremetalhost-label-sync-profile: some-ns/label-sync-profile-0
+prefixes:
+  - my-prefix.metal3.io
+  - my-prefix.kubernetes.io
+label-sync-interval: "30s"  
+```
+
+It's the admin's/user's responsibility to create the linkage by labeling BMHs (e.g `metal3.io/baremetalhost-label-sync-profile: some-ns/label-sync-profile-0`). 
+
+### Metal3Cluster 
+
+We can have the prefixes specified at the Metal3Cluster level? For example, 
+
+```
+kind: Metal3Cluster
+name: test1
+spec:
+  bmh-label-sync-prefixes: "my-prefix.metal3.io, my-prefix.kubernetes.io"
+  bmh-label-sync-interval: "30s"   
+...
+```
+
+We need to map the BMH to the Metal3Cluster object. Similarly, when Metal3Cluster is updated, we need to fetch all associated BMHs. We can investigate this further if this is a better approach. 
+
+## Other Issues
+
+### Delay between Node create and label sync
+
+The Node object is first created when kubeadm joins a node to the workload cluster (i.e. kubelet is up and running). There may be a delay (potentially several seconds) before the CAPM3 node synchronization logic kicks in to apply the labels on the Node. This may not be of concern to most users. However, it's worth asking if we can reduce or eliminate this delay?
+
+One way to address this is to use kubelet's `--node-labels` flag. In CAPI, we can potentially utilize `kubeletExtraArgs` within the KubeadmConfig spec for this purpose. However, this would require further investigation.  In particular, since the KubeadmConfig objects are generally created from KubeadmConfigTemplate, can we dynamically patch the objects with the labels from the BMH? 
```

Comment:
There is already a similar mechanism in place in Metal3DataTemplate where we refer to a single label, we could extend it to cover any labels that match the prefix

Reply:
Thanks. I have noted that.
It also seems to me that there may be user-level mechanisms that can get around this. I have added that to the section. 
In general, I think this is something we have to see if users actually have an issue with. My preference would be to just document this for the time being. If the need arises we can take it up later.